### PR TITLE
Allow info content to support json

### PR DESCRIPTION
**Pull request by rowanwins**
### What this PR does

Related to #4612 

So in `master` we had a `Metadata` class which could contain all weird and wonderful things. In `next` we haven't ported that and are largely reliant on `InfoSection`'s which just have a content field which only accepts a `string`.

In master with Metadata we could have JSON objects which allowed us to support more complex objects such as data source stuff from a WMS
Screen_2020_08_28

So there are few options
1. Re-implement the `Metadata` class from `master` in `next`
2. Extend the `InfoSection` to support JSON objects
  - The second approach feels like it could potentially simplify things by consolidating places where info and metadata is stored. 

This PR is a basic start on that second approach. If we agreed that this was a helpful way forward we could finish tidying it up.

Being able to reference JSON metadata is a requirement for Kaleidoscope as they want to have a button in the preview section related to the layer metadata from a WMS. So this PR also includes a exposing the info array as an object, which perhaps isn't ideal, but it works.

````
       "info": [
         {
           "name": "",
           "content": "{{#infoAsObject.DataDescription}}<a href='{{infoAsObject.DataDescription.MetadataURL.OnlineResource.xlink:href}}'><button class='tjs-_buttons__btn tjs-_buttons__btn-primary'>Find metadata in Marlin</button></a>{{/infoAsObject.DataDescription}}{{^infoAsObject.DataDescription}}<p>No Marlin Metadata recorded</p>{{/infoAsObject.DataDescription}}"
          }
       ]
````

### Checklist

-   [ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Comment by rowanwins** [comment]
Here is what the result looks like from the deployment
InNext

File: lib/Traits/CatalogMemberTraits.ts [hunk 1]
```diff
@@ -13,18 +15,16 @@ export class InfoSectionTraits extends ModelTraits {
...
-    type: "string",
+  @anyTrait({
     name: "Content",
     description:
-      "The content of the section, in Markdown and HTML format. Set this property to null to remove this section entirely.",
-    isNullable: true
+      "The content of the section, this either accepts a string in Markdown and HTML format, or a JSON object, which will be rendered as a table. Set this property to null to remove this section entirely."
   })
-  content?: string | null;
+  content?: string | JsonObject | null;
```
Comment (rowanwins) [inline lib/Traits/CatalogMemberTraits.ts]: This is the key change

File: lib/ModelMixins/CatalogMemberMixin.ts [hunk 2]
```diff
@@ -78,6 +78,17 @@ function CatalogMemberMixin<T extends Constructor<CatalogMember>>(Base: T) {
       );
     }
 
+    @computed
+    get infoAsObject() {
```
Comment (rowanwins) [inline lib/ModelMixins/CatalogMemberMixin.ts:83]: This supports accessing infoSections within the mustache templates for feature info

**Comment by soyarsauce** [comment]
It seems better if this is another trait, rather than clobbering the `InfoSectionTraits`'s content trait - that way we could render both the json editor as well as a markdown editor (ignoring the fact we have no way to signal markdown-parsability of a trait in code yet)

image

**Comment by rowanwins** [comment]
Just dropping a note that I've started refactoring so that there is a new trait called `contentAsObject` rather than mixing types.

**Review by tephenavies — CHANGES_REQUESTED** [review]

File: lib/ModelMixins/CatalogMemberMixin.ts [hunk 3]
```diff
@@ -78,6 +79,24 @@ function CatalogMemberMixin<T extends Constructor<CatalogMember>>(Base: T) {
...
+      const infoObject: any = {};
+
+      this.info.forEach(infoItem => {
+        if (infoItem.name !== undefined && infoItem.name.length > 0) {
+          const infoNameNoSpaces = infoItem.name.replace(/ /g, "");
+          if (infoItem.content !== undefined) {
+            infoObject[infoNameNoSpaces] = infoItem.content;
+          } else if (infoItem.contentAsObject !== undefined) {
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/ModelMixins/CatalogMemberMixin.ts]: Not sure I follow here @steve9164... are you suggesting that where we have 2 info items that have the same `name` the one that has `contentAsObject` ought to be what is returned?

[the remark this answers is not in the document]
Comment (tephenavies) [reply lib/ModelMixins/CatalogMemberMixin.ts]: That is what I'm suggesting. Is that something we can/should do here?

[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/ModelMixins/CatalogMemberMixin.ts]: Have now amended, I think that makes sense

**Review by tephenavies — DISMISSED** [review]

**Comment by tephenavies** [comment]
Missing a CHANGES.md entry

**Review by tephenavies — CHANGES_REQUESTED** [review]
@rowanwins Please add a few lines to CHANGES.md about the various new features and add a test where the service description comes through as an info item with `contentAsObject`. Any additional tests you can think of would also be great.

**Review by tephenavies — APPROVED** [review]

**Pull request merged**